core/src/main/scala/cr/pulsar/Consumer.scala

```diff
@@ -76,16 +76,17 @@ object Consumer {
   }
 
   private def mkConsumer[
+      G[_]: Concurrent: ContextShift,
```

Comment:
Hmm I don't really see the benefits of doing this when both effects have the same (or more or less) constraints. This introduces a burden on the maintainers instead. Have you got an example that would illustrate how this might be useful to the users?

Reply:
Having more or less the same constraints is the side-effect of us using `FutureLift` which requires both `Concurrent` and `ContextShift`. We need `FutureLift` for both effects because we are using `subscribeAsync` function. Actually, maybe using it doesn't even bring that much benefit. This should be checked and then can be substituted by sync `subscribe`, like it is done in `Reader`. Then constraints for creation effect type could be reduced to just `Sync` leaving all the other methods async.

Benefits? On the user side, you can have different effect types for `F` and `G`. `G` would be used only to create the resources but `F` could be used in the runtime. For example, `F` could have a context. This context could contain data about message sources or tracing data.

Reply:
The `F` that might require context can always lift `G` into `F`, like `Resource.liftF` and so on. I get it when you have tracing and so on but I fail to see the benefit in this library. Even if the constraints are only up to `Concurrent` in the current hierarchy, you have instances for `Kleisli`, which is the default transformer that is context-aware.

Reply:
The issue appears when you are trying to lift `F` into `G` - you need to provide a context which you don't have. 
I've prepared a simplified example that is using `ZIO` to carry context, but that could be also Tofu, cats-mtl or any other context supporting effect type. 

Example A. Both `Consumer` and `Producer` support 2 different effects:
https://scastie.scala-lang.org/psisoyev/TKpXu07RQNefKCQdV3eA5g/6
Here we simply provide the context whenever we have it. 

Example B. Both `Consumer` and `Producer` are created within the same effect. 
Business logic isn't changed - it uses context. 
Whenever we try to run the `result` we have to provide some context which we don't have. 

How would you solve it? 

Regarding the burden of maintenance: we have quite a large amount of constructors, which could be replaced by one (or a few) with default parameters. However, that's a topic for another discussion 😄

Reply:
Counter-example for Example B, using Cats :wink: https://scastie.scala-lang.org/Vz4LqHw4SnWLGFSKSP4lew

I think it's a mistake to create the `Consumer` and `Producer` with an effect that requires context, why would you need that?